### PR TITLE
Metrics

### PR DESCRIPTION
**Pull request by dhuber23b**
Everything is done with the following exceptions:
- Need to integrate the perceptual metric code from VA Tech
- Voxels in the test data include internal voxels.  The white paper defines comparison only on surface voxels.  This is due to the way binvox works.  We will need to write a tool to extract only external voxels or use a different voxel generation method.
- Voxels do not currently support color evaluation (i.e., appearance metric).  The previous voxelization code was replaced by binvox (external to the codebase) and the VoxelGrid class.  I cannot tell whether the old version supported color or not -- the colors of sampled points were passed into the C++ code, but it did not seem to compute the average color (which is the presumed definition).  We will need to refactor the VoxelGrid class to support color.  It is non-trivial because the add_points routine would need to be written in C++ to make it efficient, as the union trick would not work with any voxel storage other than Boolean occupancy.

### CODE REVIEW

**Comment by lynetcha** [comment]
I will review this in more details but high-level comment - The previous version supported both surface voxel IOU and appearance evaluation, and was doing color averaging. Since it relied on sampling from the mesh primarily, it also only evaluated external voxels (as long as the input mesh only has external faces).

File: sumo-api/sumo/metrics/bb_evaluator.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+        Return:
+        float - bounding box IoU (Equation 1 in SUMO white paper)
+        """
+        box1 = _bbox2pymesh(element1)
+        box2 = _bbox2pymesh(element2)
+        inter = pymesh.boolean(box1, box2, operation='intersection', engine='cgal')
+        ivert, ifaces, _ = remove_duplicated_vertices_raw(inter.vertices, inter.faces)
+        inter_mesh = pymesh.form_mesh(ivert, ifaces)
+        # ::: Remove debugging code
```
Comment (NCBee) [inline sumo-api/sumo/metrics/bb_evaluator.py]: Remove

File: sumo-api/sumo/metrics/mesh_evaluator.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    A, B, C = faces[0::3, :], faces[1::3, :], faces[2::3, :]
+    cross = np.cross(A[:, 0:3] - C[:, 0:3] , B[:, 0:3] - C[:, 0:3])
+    areas = 0.5 * (np.sqrt(np.sum(cross**2, axis=1)))
+
+    # ::: set minimum of 1 sample?  ::: this will fail if all faces
+    # are small enough
+    Nsamples_per_face = (density*areas).astype(int)
+    N = np.sum(Nsamples_per_face)  # N = total # of samples
+
+    #TODO: Remove after mesh reading bug is fixed :::? what?
```
Comment (NCBee) [inline sumo-api/sumo/metrics/mesh_evaluator.py]: Revisit

File: sumo-api/sumo/metrics/tests/voxel_evaluator_tests.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+        rms_points_error = evaluator.rms_points_error()
+        self.assertTrue(rms_points_error < 0.05)
+
+    @unittest.skip
+    def test_rms_color_error(self):
+        evaluator = VoxelEvaluator(self.submission, self.ground_truth, self.settings)
+
+        rms_color_error = evaluator.rms_color_error()
+        print(rms_color_error)
```
Comment (NCBee) [inline sumo-api/sumo/metrics/tests/voxel_evaluator_tests.py]: Remove

File: sumo-api/sumo/metrics/voxel_evaluator.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+        Inputs:
+        submission (ProjectScene) - Submitted scene to be evaluated
+        ground_truth (ProjectScene) - The ground truth scene
+        settings (EasyDict) - configuration for the evaluator.  See
+        Evaluator.py for recognized keys and values.   
+        """
+
+
+        # TODO: not needed anymore.  Use this as a template for converter
```
Comment (NCBee) [inline sumo-api/sumo/metrics/voxel_evaluator.py]: If not needed, remove

File: sumo-api/sumo/metrics/tests/mesh_evaluator_tests.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        
+    def test_shape_similarity(self):
+        """
+        Verify that the shape similarity measure is producing sane outputs.
+        """
+        
+        # make a dummy scene
+        scene = ProjectScene("meshes")
+
+        # TODO: Get deepcopy working for ProjectScene and make a simpler example for faster unit test.
```
Comment (NCBee) [inline sumo-api/sumo/metrics/tests/mesh_evaluator_tests.py]: Is this dependent on GltfModel deep copy and has it been addressed?

File: sumo-api/sumo/metrics/tests/voxel_evaluator_tests.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    @unittest.skip
+    def test_shape_similarity(self):
+        """
+        Verify that the shape similarity measure is producing sane outputs.
+        """
+        
+        # make a dummy scene
+        scene = ProjectScene("voxels")
+
+        # TODO: Get deepcopy working for ProjectScene and make a simpler example for faster unit test.
```
Comment (NCBee) [inline sumo-api/sumo/metrics/tests/voxel_evaluator_tests.py]: Is this resolved?

File: sumo-api/sumo/metrics/tests/bb_evaluator_tests.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+                               "Expected semantic score of %.3f, found %.3f.\n" %
+                               (expected, semantic_score))
+        
+        
+        
+        
+        
+
+
+    # def test_association(self):
```
Comment (NCBee) [inline sumo-api/sumo/metrics/tests/bb_evaluator_tests.py]: Remove commented code

File: sumo-api/sumo/metrics/tests/util_tests.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+from sumo.geometry.rot3 import Rot3
+import sumo.metrics.utils as utils
+
+"""
+    Test Evaluator utils functions
+"""
+
+
+
+# def create_box(center, w, h,l, anglexyz):
```
Comment (NCBee) [inline sumo-api/sumo/metrics/tests/util_tests.py]: Remove commented code

File: sumo-api/sumo/metrics/tests/util_tests.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        euler = [rz, ry, rx]
+        m = utils.euler_to_matrix(euler)
+        euler2 = utils.matrix_to_euler(m)
+
+        for i in range(3):
+            self.assertAlmostEqual(euler[i], euler2[i]) 
+    
+
+
+    # def test_iou(self):
```
Comment (NCBee) [inline sumo-api/sumo/metrics/tests/util_tests.py]: Remove commented code.

File: sumo-api/sumo/metrics/voxel_evaluator.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        # TODO: Add check that scene type is voxels
+        
+        # # sample GT, voxelize, and store in voxels field
```
Comment (NCBee) [inline sumo-api/sumo/metrics/voxel_evaluator.py]: Remove commented code

File: sumo-api/sumo/metrics/voxel_evaluator.py (hunk 4 above)
Comment (dhuber23b) [reply sumo-api/sumo/metrics/voxel_evaluator.py]: We will use this in the code shortly.  Prefer to leave it in for now.

File: sumo-api/sumo/metrics/tests/mesh_evaluator_tests.py (hunk 5 above)
Comment (dhuber23b) [reply sumo-api/sumo/metrics/tests/mesh_evaluator_tests.py]: We have a deepcopy, but it does not work like we need it to to be able to make copies of ProjectScenes

File: sumo-api/sumo/metrics/tests/voxel_evaluator_tests.py (hunk 6 above)
Comment (dhuber23b) [reply sumo-api/sumo/metrics/tests/voxel_evaluator_tests.py]: No.

**Pull request merged**